crates/chain/src/keychain/txout_index.rs

```diff
@@ -559,32 +555,107 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
     }
 }
 
-fn range_descriptor_spks<'a, R>(
-    descriptor: Cow<'a, Descriptor<DescriptorPublicKey>>,
-    range: R,
-) -> impl Iterator<Item = (u32, Script)> + Clone + Send + 'a
+/// An iterator for derived script pubkeys.
+///
+/// [`SpkIterator`] is an implementation of the [`Iterator`] trait which possesses its own `next()`
+/// and `nth()` functions, both of which circumvent the unnecessary intermediate derivations required
+/// when using their default implementations.
+///
+/// ## Examples
+///
+/// ```
+/// use bdk_chain::keychain::SpkIterator;
+/// # use miniscript::{Descriptor, DescriptorPublicKey};
+/// # use bitcoin::{secp256k1::Secp256k1};
+/// # use std::str::FromStr;
+/// # let secp = bitcoin::secp256k1::Secp256k1::signing_only();
+/// # let (descriptor, _) = Descriptor::<DescriptorPublicKey>::parse_descriptor(&secp, "wpkh([73c5da0a/86'/0'/0']xprv9xgqHN7yz9MwCkxsBPN5qetuNdQSUttZNKw1dcYTV4mkaAFiBVGQziHs3NRSWMkCzvgjEe3n9xV8oYywvM8at9yRqyaZVz6TYYhX98VjsUk/1/0)").unwrap();
+///
+/// // Creates a new script pubkey iterator starting at 0 from a descriptor.
+/// let _spk_iter = SpkIterator::new(&descriptor);
+///
+/// // Creates a new script pubkey iterator from a descriptor with a given range, in this
+/// // case between 0 and a constant representing the maximum BIP32 derivation index.
+/// const BIP32_MAX_INDEX: u32 = (1 << 31) - 1;
+/// let _spk_range_iter = SpkIterator::new_with_range(&descriptor, 0..BIP32_MAX_INDEX);
+/// ```
+#[derive(Clone)]
+pub struct SpkIterator<D> {
+    next_index: u32,
+    end: u32,
+    descriptor: D,
+    secp: Secp256k1<bitcoin::secp256k1::VerifyOnly>,
+}
+
+impl<D> SpkIterator<D>
 where
-    R: Iterator<Item = u32> + Clone + Send + 'a,
+    D: Borrow<Descriptor<DescriptorPublicKey>>,
 {
-    let secp = Secp256k1::verification_only();
-    let has_wildcard = descriptor.has_wildcard();
-    range
-        .into_iter()
-        // non-wildcard descriptors can only have one derivation index (0)
-        .take_while(move |&index| has_wildcard || index == 0)
-        // we can only iterate over non-hardened indices
-        .take_while(|&index| index <= BIP32_MAX_INDEX)
-        .map(
-            move |index| -> Result<_, miniscript::descriptor::ConversionError> {
-                Ok((
-                    index,
-                    descriptor
-                        .at_derivation_index(index)
-                        .derived_descriptor(&secp)?
-                        .script_pubkey(),
-                ))
+    /// Creates a new script pubkey iterator starting at 0 from a descriptor.
+    pub fn new(descriptor: D) -> Self {
+        let end = if descriptor.borrow().has_wildcard() {
+            BIP32_MAX_INDEX
+        } else {
+            0
+        };
+
+        SpkIterator::new_with_range(descriptor, 0..=end)
+    }
+
+    /// Creates a new script pubkey iterator from a descriptor with a given range.
+    pub fn new_with_range<R>(descriptor: D, range: R) -> Self
```

Comment:
I don't think we need `new_with_range`. You can just `.skip` to the start of the range and `.take` the length of the range you want.

Reply:
That is a good point. So we don't even need the `end` field of `SpkIterator`.

Reply:
However, I would prefer that we keep this method, just redefine it to use `.skip` and `.take`. This is a helpful method to use imo.

Reply:
@LLFourn Are you concerned because we are using the `RangeBounds` trait in a "weird" way?

Reply:
I just don't know why we want to create a method that just calls `.skip` and `.take` when the caller can just do this. We don't have to document it or test it then.

Reply:
Please refer to https://github.com/bitcoindevkit/bdk/pull/927#issuecomment-1527938350